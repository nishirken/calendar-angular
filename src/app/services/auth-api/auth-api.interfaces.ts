import { ApiError } from '../api.interfaces';

export enum SigninErrorCode {
  CredsInvalid = 'CREDENTIALS_INVALID',
}

export type SigninApiError = ApiError<SigninErrorCode>;

export enum SignupErrorCode {
  EmailInvalid = 'EMAIL_INVALID',
  UserExists = 'USER_EXISTS',
  PasswordInvalid = 'PASSWORD_INVALID',
}

export enum PasswordInvalidErrorCode {
  TooShort = 'PASSWORD_TOO_SHORT',
  TooLong = 'PASSWORD_TOO_LONG',
  ReqChars = 'NOT_ENOUGH_REQ_CHARS',
  InvalidChars = 'INVALID_CHARS',
}

export type PasswordInvalidError =
  | {
      code: PasswordInvalidErrorCode.TooShort;
      minLength: number;
      providedLength: number;
    }
  | {
      code: PasswordInvalidErrorCode.TooLong;
      maxLength: number;
      providedLength: number;
    }
  | {
      code: PasswordInvalidErrorCode.ReqChars;
      minimumAmount: number;
      providedAmount: number;
      characterCategory: string;
    }
  | {
      code: PasswordInvalidErrorCode.InvalidChars;
      invalidCharacters: string;
    };

export type SignupApiError =
  | ApiError<SignupErrorCode.EmailInvalid, { reason: string }>
  | ApiError<SignupErrorCode.UserExists>
  | ApiError<SignupErrorCode.PasswordInvalid, PasswordInvalidError[]>;

export type AuthCreds = {
  email: string;
  password: string;
};
