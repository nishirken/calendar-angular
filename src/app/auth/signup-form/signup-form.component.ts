import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { signupFormTestIds } from 'calendar-test-ids';
import { RoutePath } from '../../app-routing.module';
import {
  AuthCreds,
  PasswordInvalidErrorCode,
  SignupApiError,
  SignupErrorCode,
} from '../../services/auth-api/auth-api.interfaces';
import { AuthApiService } from '../../services/auth-api/auth-api.service';
import { AuthFormComponent } from '../auth-form/auth-form.component';

@Component({
  selector: 'app-signup-form',
  templateUrl: '../auth-form/auth-form.component.html',
  styleUrls: ['../auth-form/auth-form.component.sass'],
})
export class SignupFormComponent extends AuthFormComponent {
  constructor(private authApiService: AuthApiService, private router: Router) {
    super();
  }

  override buttonText = 'Sign up';
  override hintText = 'Already have an account?';
  override hintLinkText = 'Sign in';
  override hintLinkHref = '/auth/signin';
  override testIds = signupFormTestIds;

  override submit = (creds: AuthCreds) => {
    this.authApiService.signup(creds).subscribe({
      next: () => this.router.navigate([RoutePath.Week]),
      error: this.handleError,
    });
  };

  handleError = (error: SignupApiError) => {
    switch (error.error.code) {
      case SignupErrorCode.EmailInvalid:
        this.emailErrorMessage = `Invalid: ${error.error.data.reason}`;
        break;
      case SignupErrorCode.UserExists:
        this.emailErrorMessage = 'This email already exists';
        break;
      case SignupErrorCode.PasswordInvalid:
        const firstError = error.error.data[0];
        if (!firstError) {
          return;
        }
        switch (firstError.code) {
          case PasswordInvalidErrorCode.TooShort:
            this.passwordErrorMessage = `Must have at least ${firstError.minLength} symbols`;
            break;
          case PasswordInvalidErrorCode.TooLong:
            this.passwordErrorMessage = `Must have max ${firstError.maxLength} sybmols`;
            break;
          case PasswordInvalidErrorCode.ReqChars:
            console.log(error.error);
            this.passwordErrorMessage = `Missing ${firstError.minimumAmount} ${firstError.characterCategory} symbols`;
            break;
          case PasswordInvalidErrorCode.InvalidChars:
            this.passwordErrorMessage = `Invalid symbols ${firstError.invalidCharacters}`;
        }
    }
  };
}
