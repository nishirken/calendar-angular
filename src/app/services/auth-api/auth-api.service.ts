import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, switchMap } from 'rxjs';
import { url } from '../helpers';
import { AuthCreds } from './auth-api.interfaces';

@Injectable({
  providedIn: 'root',
})
export class AuthApiService {
  constructor(private http: HttpClient) {}

  signin = (creds: AuthCreds): Observable<void> => {
    return this.http.post<void>(url('/auth/signin'), creds, {
      withCredentials: true,
    });
  };

  signup = (creds: AuthCreds): Observable<void> => {
    return this.http
      .post<void>(url('/auth/signup'), creds, { withCredentials: true })
      .pipe(switchMap(() => this.signin(creds)));
  };
}
