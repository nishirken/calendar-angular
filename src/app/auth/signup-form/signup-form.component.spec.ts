import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AuthApiService } from '../../services/auth-api/auth-api.service';
import { AuthApiService as AuthApiServiceStub } from '../../services/auth-api/__mocks__/auth-api.service';

import { SignupFormComponent } from './signup-form.component';

describe('SignupFormComponent', () => {
  let component: SignupFormComponent;
  let fixture: ComponentFixture<SignupFormComponent>;
  const authApiService = new AuthApiServiceStub();

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [SignupFormComponent],
      providers: [{ provide: AuthApiService, useValue: AuthApiServiceStub }],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents();

    fixture = TestBed.createComponent(SignupFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(fixture).toMatchSnapshot();
  });
});
