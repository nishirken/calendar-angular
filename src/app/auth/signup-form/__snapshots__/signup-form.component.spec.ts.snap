// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupFormComponent should create 1`] = `
<app-signup-form
  __ngContext__="0"
  authApiService={[Function Function]}
  buttonText={[Function String]}
  emailControl={[Function FormControl2]}
  emailErrorMessage=""
  handleError={[Function Function]}
  handleLinkClick={[Function Function]}
  handleSubmit={[Function Function]}
  hintLinkHref={[Function String]}
  hintLinkText={[Function String]}
  hintText={[Function String]}
  href={[Function String]}
  passwordControl={[Function FormControl2]}
  passwordErrorMessage=""
  router={[Function Router]}
  submit={[Function Function]}
  testIds={[Function Object]}
>
  <div
    class="form"
    data-test-id="signup-form"
  >
    <app-input
      class="input"
      placeholder="email"
      type="email"
    />
    <app-input
      class="input"
      placeholder="password"
      type="password"
    />
    <wired-button
      class="button"
      data-test-id="signup-form-submit-button"
    >
       Sign up 
    </wired-button>
    <div
      class="hint"
    >
       Already have an account? 
      <wired-link
        class="hint-link"
        data-test-id="signup-form-link"
      >
         Sign in 
      </wired-link>
    </div>
  </div>
</app-signup-form>
`;
